module/data/fields.mjs

```diff
@@ -0,0 +1,95 @@
+/**
+ * Data field that automatically selects the Advancement-specific configuration or value data models.
+ *
+ * @param {Advancement} advancementType  Advancement class to which this field belongs.
+ */
+export class AdvancementDataField extends foundry.data.fields.ObjectField {
+  constructor(advancementType, options={}) {
+    super(options);
+    this.advancementType = advancementType;
+  }
+
+  /** @inheritdoc */
+  static get _defaults() {
+    return foundry.utils.mergeObject(super._defaults, {required: true});
+  }
+
+  getModel() {
+    return this.advancementType.metadata?.dataModels?.[this.name];
+  }
+
+  getDefaults() {
+    return this.advancementType.metadata?.defaults?.[this.name] ?? {};
+  }
+
+  _cleanType(value, options) {
+    if ( !(typeof value === "object") ) value = {};
+
+    // Use a defined DataModel
+    const cls = this.getModel();
+    if ( cls ) return cls.cleanData(value, options);
+    if ( options.partial ) return value;
+
+    // Use the defined defaults
+    const defaults = this.getDefaults();
+    return foundry.utils.mergeObject(defaults, value, {inplace: true});
+  }
+
+  initialize(value, model) {
+    const cls = this.getModel();
+    if ( cls ) return new cls(value, {parent: model});
+    return foundry.utils.deepClone(value);
+  }
+}
+
+/* -------------------------------------------- */
+
+/**
+ * @typedef {StringFieldOptions} FormulaFieldOptions
+ * @property {boolean} [deterministic=false]  Is this formula not allowed to have dice values?
+ */
+
+/**
+ * Special case StringField which represents a formula.
+ *
+ * @param {FormulaFieldOptions} [options={}]  Options which configure the behavior of the field.
+ * @property {boolean} deterministic=false    Is this formula not allowed to have dice values?
+ */
+export class FormulaField extends foundry.data.fields.StringField {
+
+  /** @inheritdoc */
+  static get _defaults() {
+    return foundry.utils.mergeObject(super._defaults, {
+      deterministic: false
+    });
+  }
+
+  /** @inheritdoc */
+  _validateType(value) {
+    const roll = new Roll(value);
+    roll.evaluate({async: false});
```

Comment:
Yes, I needed a way to validate a roll that would provide a proper error message. `Roll.validate` unfortunately catches all errors and only returns `false`, so I simply did what it does without the try…catch block.

Reply:
What would be the consequences of not validating these rolls? Would they just fail later when execution is attempted?

Reply:
Yep, presumably. Just the question of whether we want to fail early when the user enters the data or later when they try to use it.

Maybe we could leave it in now and add a request for a roll validation method in core that returns errors properly for when the `async: false` option is removed entirely from `Roll`?

Reply:
Added an additional configuration option to `FormulaField` to properly support the max uses field (cause it uses `Roll#safeEval` to evaluate formulas, not `Roll#evaluate`).

Reply:
It's probably possible to have `Roll.validate` throw its error rather than returning `false`, but I'm not really sure at this stage how `Roll.validate` is even going to work with `async: false` eventually removed. I don't feel strongly here so it's fine to leave it for now if you prefer.